src/Features/Core/Portable/GenerateMember/GenerateConstructor/AbstractGenerateConstructorService.cs

```diff
@@ -51,21 +85,17 @@ public async Task<ImmutableArray<CodeAction>> GenerateConstructorAsync(Document
                 var state = await State.GenerateAsync((TService)this, semanticDocument, node, cancellationToken).ConfigureAwait(false);
                 if (state != null)
                 {
-                    var result = ArrayBuilder<CodeAction>.GetInstance();
-                    var codeAction = new GenerateConstructorCodeAction((TService)this, document, state, withFields: true);
-                    result.Add(codeAction);
-
-                    // First see the type of edit our regular code action would create.  If it 
-                    // creates fields, then also offer to perform the code action without creating
-                    // any fields.
-                    var edit = await codeAction.GetEditAsync(cancellationToken).ConfigureAwait(false);
-                    if (edit.addedFields)
-                    {
-                        result.Add(
-                            new GenerateConstructorCodeAction((TService)this, document, state, withFields: false));
-                    }
-
-                    return result.ToImmutableAndFree();
+                    using var _ = ArrayBuilder<CodeAction>.GetInstance(out var result);
+
+                    if (state.ParameterToNewFieldMap.Count > 0)
+                        result.Add(new GenerateConstructorCodeAction(document, state, withFields: true, withProperties: false));
+
+                    if (state.ParameterToNewPropertyMap.Count > 0)
+                        result.Add(new GenerateConstructorCodeAction(document, state, withFields: false, withProperties: true));
+
+                    result.Add(new GenerateConstructorCodeAction(document, state, withFields: false, withProperties: false));
+
+                    return result.ToImmutable();
```

Comment:
this logic is much cleaner than before.  note how the old logic would create the code action, then apply it to see if it added fields, then based on that it would offer to generate a simple constructor that didn't do that. 

We avoid that now entirely.  no need to apply the existing code action to determine what to do.